crates/wasmi/src/limits.rs

```diff
@@ -0,0 +1,303 @@
+        self.0.memories = memories;
+        self
+    }
+
+    /// Indicates that a trap should be raised whenever a growth operation
+    /// would fail.
+    ///
+    /// This operation will force `memory.grow` and `table.grow` instructions
+    /// to raise a trap on failure instead of returning -1. This is not
+    /// necessarily spec-compliant, but it can be quite handy when debugging a
+    /// module that fails to allocate memory and might behave oddly as a result.
+    ///
+    /// This value defaults to `false`.
+    pub fn trap_on_grow_failure(mut self, trap: bool) -> Self {
+        self.0.trap_on_grow_failure = trap;
+        self
+    }
+
+    /// Consumes this builder and returns the [`StoreLimits`].
+    pub fn build(self) -> StoreLimits {
+        self.0
+    }
+}
+
+impl Default for StoreLimitsBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Provides limits for a [`Store`](crate::Store).
+///
+/// This type is created with a [`StoreLimitsBuilder`] and is typically used in
+/// conjunction with [`Store::limiter`](crate::Store::limiter).
+///
+/// This is a convenience type included to avoid needing to implement the
+/// [`ResourceLimiter`] trait if your use case fits in the static configuration
+/// that this [`StoreLimits`] provides.
+#[derive(Clone, Debug)]
+pub struct StoreLimits {
+    memory_size: Option<usize>,
+    table_elements: Option<u32>,
+    instances: usize,
+    tables: usize,
+    memories: usize,
+    trap_on_grow_failure: bool,
+}
+
+impl Default for StoreLimits {
+    fn default() -> Self {
+        Self {
+            memory_size: None,
+            table_elements: None,
+            instances: DEFAULT_INSTANCE_LIMIT,
+            tables: DEFAULT_TABLE_LIMIT,
+            memories: DEFAULT_MEMORY_LIMIT,
+            trap_on_grow_failure: false,
+        }
+    }
+}
+
+impl ResourceLimiter for StoreLimits {
+    fn memory_growing(
+        &mut self,
+        _current: usize,
+        desired: usize,
+        maximum: Option<usize>,
+    ) -> Result<bool, MemoryError> {
+        let allow = match self.memory_size {
+            Some(limit) if desired > limit => false,
+            _ => match maximum {
+                Some(max) if desired > max => false,
+                Some(_) | None => true,
+            },
+        };
+        if !allow && self.trap_on_grow_failure {
+            Err(MemoryError::OutOfBoundsGrowth)
+        } else {
+            Ok(allow)
+        }
+    }
+
+    fn table_growing(
+        &mut self,
+        current: u32,
+        desired: u32,
+        maximum: Option<u32>,
+    ) -> Result<bool, TableError> {
+        let allow = match self.table_elements {
+            Some(limit) if desired > limit => false,
+            _ => match maximum {
+                Some(max) if desired > max => false,
+                Some(_) | None => true,
+            },
+        };
+        if !allow && self.trap_on_grow_failure {
```

Comment:
Technically we could check if `self.trap_on_grow_failure` was `false` and bail out early instead, right?

Reply:
I'm not sure what you mean by early. The `allow` flag is a `bool` that has its full value -- either true or false -- returned in the false branch here. I don't think the conditional can be reorganized to be any simpler.

Reply:
Nevermind, shouldn't review when it is so late. 🙃